docs/configuration/http_filters/lua_filter.rst

```diff
@@ -0,0 +1,325 @@
+has been buffered. Note that all buffering must adhere to the flow control policies in place.
+Envoy will not buffer more data than is allowed by the connection manager.
+
+Returns a :ref:`buffer object <config_http_filters_lua_buffer_wrapper>`.
+
+bodyChunks()
+^^^^^^^^^^^^
+
+.. code-block:: lua
+
+  iterator = handle:bodyChunks()
+
+Returns an iterator that can be used to iterate through all received body chunks as they arrive.
+Envoy will yield rthe script in between chunks, but *will not buffer* them. This can be used by
+a script to inspect data as it is streaming by.
+
+.. code-block:: lua
+
+  for chunk in request_handle:bodyChunks() do
+    request_handle:log(0, chunk:byteSize())
+  end
+
+Each chunk the iterator returns is a :ref:`buffer object <config_http_filters_lua_buffer_wrapper>`.
+
+trailers()
+^^^^^^^^^^
+
+.. code-block:: lua
+
+  trailers = handle:trailers()
+
+Returns the stream's trailers. May return nil if there are no trailers. The trailers may be
+modified before they are sent to the next filter.
+
+Returns a :ref:`header object <config_http_filters_lua_header_wrapper>`.
+
+log()
+^^^^^
+
+.. code-block:: lua
+
+  handle:log(level, message)
+
+Logs a message using Envoy's application logging. *level* is an integer in the range 0 - 5 which
+maps to the spdlog levels *trace* - *critical*. *message* is a string to log.
+
+httpCall()
+^^^^^^^^^^
+
+.. code-block:: lua
+
+  headers, body = handle:httpCall(cluster, headers, body, timeout)
+
+Makes an HTTP call to an upstream host. Envoy will yield the script until the call completes or
+has an error. *cluster* is a string which maps to a configured cluster manager cluster. *headers*
+is a table of key/value pairs to send. Note that the *:method*, *:path*, and *:authority* headers
+must be set. *body* is an optional string of body data to send. *timeout* is an integer that
+specifies the call timeout in milliseconds.
+
+Returns *headers* which is a table of response headers. Returns *body* which is the string response
+body. May be nil if there is no body.
+
+respond()
+^^^^^^^^^^
+
+.. code-block:: lua
+
+  handle:respond(headers, body)
+
+Respond immediately and do not continue further filter iteration. This call is *only valid in
+the request flow*. *headers* is a table of key/value pairs to send. Note that the *:status* header
+must be set. *body* is a string and supplies the optional response body. May be nil.
+
+.. _config_http_filters_lua_header_wrapper:
+
+Header object API
+-----------------
+
+add()
+^^^^^
+
+.. code-block:: lua
+
+  headers:add(key, value)
+
+Adds a header. *key* is a string that supplies the header key. *value* is a string that supplies
+the header value.
+
+get()
+^^^^^
+
+.. code-block:: lua
+
+  headers:get(key)
+
+Gets a header. *key* is a string that suplies the header key. Returns a string that is the header
+value or nil if there is no such header.
+
+iterate()
+^^^^^^^^^
+
+.. code-block:: lua
+
+  headers:iterate(
```

Comment:
It would be nice if you could make pairs() work for the type.  See http://lua-users.org/wiki/GeneralizedPairsAndIpairs

Reply:
This has to do with the underlying header map API. Given that we have to copy to strings anyway, and perf doesn't matter as much, I can probably also supports a pairs() iterator. Let me look into it.

Reply:
FYI looks like to get generalized pairs we would need to enable `-DLUAJIT_ENABLE_LUA52COMPAT`. Are we OK with that? http://luajit.org/extensions.html

Reply:
I'm generally in favor of those 5.2 features, but I don't feel strongly about it.

Reply:
+1 for ipairs!

Reply:
doing pairs() is a little annoying without adding https://github.com/envoyproxy/envoy/issues/1851, but I can do it in a slightly hacky way that we can clean up later so at least the Lua API is consistent. Will change this up and enable `LUAJIT_ENABLE_LUA52COMPAT` and document.